enhancements/installer/gcp-ipi-shared-vpc.md

```diff
@@ -0,0 +1,260 @@
+  done manually by the user prior to running the installer.
+- Provisioning and/or configuring the Service Accounts for the Installer and
+  Cluster are to be done manually by the user prior to running the installer.
+
+## Proposal
+
+The installer should be able to install a GCP IPI cluster using Shared VPC
+networking. This requires some resources to be created and configured in
+advance, such as the network and subnets. Other resources can be created in
+advance when the necessary host project permissions are not desired, such as
+the firewall rules and/or DNS configuration. The installer, and some operators,
+will be updated to enable this functionality. They will become capable of
+creating and managing resources that exist in a different (Host) project from
+the cluster (Service Project). They will also become capable of consuming
+pre-existing firewall and/or DNS configurations.
+
+### Workflow Description
+
+Prior to executing the installer, the Shared VPC, Network, and Subnets will be
+manually created in the Host Project and configured for Shared VPC (XPN)
+configuration with the Service Project. A service account will also be created
+in the Service Project and granted sufficient permissions in the Host Project.
+When minimal permissions are desired in the Host Project, additional resources
+will be manually created and configured in the Host Project to include:
+firewall rules, api compute address, ingress compute address, and the api,
+api-int, and \*.apps DNS records.
+
+After all of these resources are created, they will be used in a manually
+created install-config.yaml to configure the cluster accordingly. At a minimum,
+the credentialsMode will be set to Passthrough, and the network,
+computeNetwork, controlPlaneNetwork, and networkProjectID will be configured to
+reference the Shared VPC network. When minimial permissions are desired in the
+Host Project, the additional parameters will be configured to enable the
+cluster to consume them as well.
+
+The installation process can then proceed as any other IPI workflow by running
+the create manifests, create ignition, and/or create cluster sub-commands.
+
+### API Extensions
+
+#### api.dns
+
+The current DNSZone struct does not appear to be able to store which project
+the zone is in. The ID variable can be used to store fully qualified path
+names, which will include the project. However, this may require modification
+of all consumers of this variable to be able to parse this information.
+Alternatively, a second variable can be added to store the projectID and the
+consumers can use that information when linking to the DNS Zone.
+
+#### api.ingress
+
+The current ingress structures do not appear to be able to store an address to
+be used when creating the ingress load balancer(s). New field(s) may be needed
+to enable this functionality. They can then be passed to the gcp lagacy
+provider for creating the ingress load balancers.
```

Comment:
Will this be by setting the information in the cloud-config?

Do we know if any additional changes are needed for the out of tree cloud provider?

Reply:
Correct. Most of what is done in the "GCP UPI XPN" workflow will be replicated here. This includes adding the network information to the `manifests/cloud-provider-config.yaml` (network-project-id, network-name, and subnetwork-name) during the 'create manifests` stage of the installer. For the private DNS zone, I believe we can assume it exists in the same network-project-id. However, for the public zone the plan is to be more flexible and allow any project, if possible.

Reply:
Sounds good, thanks for the explanation. Might be worth adding some context early in the enhancement that mentions we are implementing this based on the UPI flow and the changes specifically around the cloud provider config are also good to note, thinking about future readers such as the docs and qe teams

Reply:
Done.